src/pkg/task/execution.go

```diff
@@ -43,3 +53,211 @@ type ExecutionManager interface {
+		}
+		if err = e.taskDAO.Delete(ctx, task.ID); err != nil {
+			return err
+		}
+	}
+
+	return e.executionDAO.Delete(ctx, id)
+}
+
+func (e *executionManager) Get(ctx context.Context, id int64) (*Execution, error) {
+	execution, err := e.executionDAO.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return e.populateExecution(ctx, execution), nil
+}
+
+func (e *executionManager) List(ctx context.Context, query *q.Query) ([]*Execution, error) {
+	executions, err := e.executionDAO.List(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	var execs []*Execution
+	for _, execution := range executions {
+		execs = append(execs, e.populateExecution(ctx, execution))
+	}
+	return execs, nil
+}
+
+func (e *executionManager) populateExecution(ctx context.Context, execution *dao.Execution) *Execution {
+	exec := &Execution{
+		ID:            execution.ID,
+		VendorType:    execution.VendorType,
+		VendorID:      execution.VendorID,
+		Status:        execution.Status,
+		StatusMessage: execution.StatusMessage,
+		Metrics:       nil,
+		Trigger:       execution.Trigger,
+		StartTime:     execution.StartTime,
+		EndTime:       execution.EndTime,
+	}
+
+	if len(execution.ExtraAttrs) > 0 {
+		extras := map[string]interface{}{}
+		if err := json.Unmarshal([]byte(execution.ExtraAttrs), &extras); err != nil {
+			log.Errorf("failed to unmarshal the extra attributes of execution %d: %v", execution.ID, err)
+		} else {
+			exec.ExtraAttrs = extras
+		}
+	}
+
+	// if the status isn't null which means the status is set manually, return directly
+	if len(exec.Status) > 0 {
+		return exec
+	}
+
+	// populate task metrics
+	e.populateExecutionMetrics(ctx, exec)
+	// populate status
+	e.populateExecutionStatus(exec)
+	// populate the end time
+	e.populateExecutionEndTime(ctx, exec)
+
+	return exec
+}
+
+func (e *executionManager) populateExecutionMetrics(ctx context.Context, execution *Execution) {
+	scs, err := e.taskDAO.ListStatusCount(ctx, execution.ID)
+	if err != nil {
+		log.Errorf("failed to list status count of execution %d: %v", execution.ID, err)
+		return
+	}
+
+	metrics := &Metrics{}
+	for _, sc := range scs {
+		switch sc.Status {
+		case job.SuccessStatus.String():
+			metrics.SuccessTaskCount = sc.Count
+		case job.ErrorStatus.String():
+			metrics.ErrorTaskCount = sc.Count
+		case job.PendingStatus.String():
+			metrics.PendingTaskCount = sc.Count
+		case job.RunningStatus.String():
+			metrics.RunningTaskCount = sc.Count
+		case job.ScheduledStatus.String():
+			metrics.ScheduledTaskCount = sc.Count
+		case job.StoppedStatus.String():
+			metrics.StoppedTaskCount = sc.Count
+		}
+	}
+	metrics.TaskCount = metrics.SuccessTaskCount + metrics.ErrorTaskCount +
+		metrics.PendingTaskCount + metrics.RunningTaskCount +
+		metrics.ScheduledTaskCount + metrics.StoppedTaskCount
+	execution.Metrics = metrics
+}
+
+func (e *executionManager) populateExecutionStatus(execution *Execution) {
+	metrics := execution.Metrics
+	if metrics == nil {
+		return
```

Comment:
What's the default status?

Reply:
No default status. If the execution contains any tasks, the status is calculated from the task statuses, or the status should be set by caller explicitly.

Reply:
By your design, there should not be an execution without any tasks, right? Shall we make sure this rule by adding validation logic?

Reply:
There are indeed executions contain no tasks. e.g. starting an execution for replication will create an execution record first, and after the initial/prepare/filter works, there maybe no artifact candidates to replicate, the execution contains no tasks in this case and the caller should set the status and end time for the execution record